nightly-job/sanitation_job.py

```diff
@@ -32,30 +32,33 @@ async def run_sanitation(args):
         
         data_validation_sample = pd.DataFrame()
 
-        unsanitized_search_term_stream = stream_search_terms()  # load unsanitized search terms
-        
+        unsanitized_search_term_stream = stream_search_terms() # load unsanitized search terms
         for raw_page in unsanitized_search_term_stream:
             total_run += raw_page.shape[0]
-            
+        
             one_percent_sample = raw_page.sample(frac = 0.01)
             data_validation_sample = data_validation_sample.append(one_percent_sample)
-            
-            pii_in_query_mask, run_data, language_data = await detect_pii(raw_page['query'], census_surnames)
-            sanitized_page = raw_page.loc[
-                ~numpy.array(pii_in_query_mask)]  # ~ reverses the mask so we get the queries WITHOUT PII in them
-            total_deemed_sanitary += sanitized_page.shape[0]
+        
+            allow_listed_terms_page = raw_page.loc[raw_page.present_in_allow_list]
+            unsanitized_unallowlisted_terms = raw_page.loc[~raw_page.present_in_allow_list]
 
+            pii_in_query_mask, run_data, language_data = await detect_pii(unsanitized_unallowlisted_terms['query'], census_surnames)
+            sanitized_page = unsanitized_unallowlisted_terms.loc[~numpy.array(pii_in_query_mask)] # ~ reverses the mask so we get the queries WITHOUT PII in them
+            total_deemed_sanitary += sanitized_page.shape[0]
```

Comment:
Here `total_deemed_sanitary` does not include the allow-listed terms. However, then `total_rejected=total_run - total_deemed_sanitary` _does_ include allow-listed terms (I think this explains @rebecca-burwei 's point 3.)
This will feed into the `pct_sanitized_search_terms` validation metric. 

I think this is probably undesirable, since then that validation metric would track changes to the allowlist and could generate false alarms on that, but open to other thoughts (cc @rebecca-burwei as well). Either way, it would be good to document that the "rejected" count includes this new group.

Alternatively, we could include the allow-listed terms in `total_deemed_sanitary`, so that `total_rejected` is unaffected. 

An additional consideration is whether the count of processed allow-listed terms is computable from the metadata (nice to have counts of the different cases broken out for sanity checking). With the current version, it is computable as `total_rejected - num_@ - num_numeric - num_name`. If allow-listed are counted in `total_deemed_sanitary`, the count of allow-listed is no longer recoverable from the metadata. In that case it would be nice to have the allow-listed count broken out as a separate column in the metadata, but I'd be ok leaving that till later as it involves updating the table schema.

Reply:
I think `total_deemed_sanitary` should include the allow-listed terms. 

---------

> Here `total_deemed_sanitary` does not include the allow-listed terms. However, then `total_rejected=total_run - total_deemed_sanitary` does include allow-listed terms (I think this explains @rebecca-burwei 's point 3.)
This will feed into the `pct_sanitized_search_terms` validation metric.

I'm not able to validate this statement using `prototype_sanitization_job_metadata` because there are so many nulls in the table... Maybe this is another issue?

---------
> An additional consideration is whether the count of processed allow-listed terms is computable from the metadata (nice to have counts of the different cases broken out for sanity checking). With the current version, it is computable as `total_rejected - num_@ - num_numeric - num_name`.

Not quite. You could have a search term that has both numbers and an @ symbol. In that case, the formula above would double subtract that search term, so the formula would give too low a number for the count of allow-listed search terms.

---------
Should we have a separate column for count of allow-listed search terms?

Reasons yes:
- Seems useful based on this conversation and the amount of reasoning we're having to do about it
- 19% of search terms appear in the allow list. that's a lot.

Reasons no:
- Historically, the allow-list doesn't change very often (once every 6 months to a year). I don't expect that cadence to increase until end of H1 2023 (when the Merino team is slated to have improved the backend they use to update the allow-list).

Reply:
> I'm not able to validate this statement using prototype_sanitization_job_metadata because there are so many nulls in the table... Maybe this is another issue?

Maybe - this is a guess going off the fact that the metadata shows 33% removed rather than the expected 15%.

Reply:
@dzeber is right: for now, we have a more or less arbitrary hierarchy of rejection reasons, and the rejections reason for a given row is the first one it hit. So no row appears in two of those counts, which has its own drawbacks, but means that if you add up the three rejection reason columns so far you get the total rejected!

Reply:
I think this brings up another good question: How do we want to handle changes in metrics in the future if our sanitization strategy changes? Namely, here we're adding a reason for sanitizing and we have to change the table schema. How often is that likely to happen? Often enough that I should think about making it easier to do?

Reply:
Worth noting that this table will never be enormous: it adds one record per day, maaaybe two if a job fails. We don't hit a thousand records for almost three years. So having some sort of column with JSON in it that we unpack daily is pretty reasonable.

Reply:
> I think this brings up another good question: How do we want to handle changes in metrics in the future if our sanitization strategy changes? Namely, here we're adding a reason for sanitizing and we have to change the table schema. How often is that likely to happen? Often enough that I should think about making it easier to do?

I would say not at this point. The main reasons I can think of for changing the sanitization strategy are: (1) something about the incoming search terms changes which makes our strategy less effective than we want, or (2) we want to a more granular strategy which reduces the number of terms removed. I don't see either happening more than once twice a year over the next couple of years.

Reply:
For now I'm gonna add the column and make a PR for that!

Reply:
I think there's another reason that's more likely -- something changes in product or the business / partnerships that means we have to add another reason. I can already think of another one: "removed because the terminal search term contained a name" (re: this [conversation](https://docs.google.com/document/d/1900UNss4F-8K1Nr3F4IvfzUkjkni9oStEQKOFp0PfXU/edit#bookmark=id.d1khkuhui37i)). I wouldn't count it out that we have to add more reasons like once a quarter.

Reply:
@rebecca-burwei that makes sense to me. 

It turns out (I have recently learned) we don't have to do a change request for adding columns, like I figured we did. So I think even at once per quarter, adding a column won't be that bad!